File: packages/sandboxd/pkg/server/process_test.go
// Copyright 2026 The Kubernetes Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package server

import (
	"context"
	"testing"
	"time"

	"google.golang.org/grpc"

	"sigs.k8s.io/agent-sandbox/packages/sandboxd/pkg/processmanager"
	processv1 "sigs.k8s.io/agent-sandbox/packages/sandboxd/spec/process/v1"
)

type mockStartServer struct {
	grpc.ServerStream
	ctx      context.Context
	events   []*processv1.StartResponse
}

func (m *mockStartServer) Context() context.Context {
	if m.ctx != nil {
		return m.ctx
	}
	return context.Background()
}

func (m *mockStartServer) Send(res *processv1.StartResponse) error {
	m.events = append(m.events, res)
	return nil
}

func TestProcessServerExecute(t *testing.T) {
	tempDir := t.TempDir()
	registry := processmanager.NewProcessRegistry()
	procServer := NewProcessServer(tempDir, registry)
	ctx := context.Background()

	res, err := procServer.Execute(ctx, &processv1.ExecuteRequest{
		Config: &processv1.ProcessConfig{
			Command: []string{"echo", "hello sandboxd"},
		},
	})
	if err != nil {
		t.Fatalf("Execute failed: %v", err)
	}

	if res.GetExitCode() != 0 {
		t.Errorf("expected exit code 0, got %d", res.GetExitCode())
	}

	if string(res.GetStdout()) != "hello sandboxd\n" {
		t.Errorf("expected stdout %q, got %q", "hello sandboxd\n", string(res.GetStdout()))
	}
}

func TestProcessServerStart(t *testing.T) {
	tempDir := t.TempDir()
	registry := processmanager.NewProcessRegistry()
	procServer := NewProcessServer(tempDir, registry)

	startStream := &mockStartServer{}
	err := procServer.Start(&processv1.StartRequest{
		Config: &processv1.ProcessConfig{
			Command: []string{"echo", "streaming output"},
		},
	}, startStream)

	if err != nil {
		t.Fatalf("Start failed: %v", err)
	}

	if len(startStream.events) < 3 {
		t.Fatalf("expected at least 3 events (Init, Stdout, Exit), got %d", len(startStream.events))
	}

	// Verify Init event
	initEvent := startStream.events[0].GetInit()
	if initEvent == nil || initEvent.GetProcessId() <= 0 {
		t.Errorf("expected valid InitEvent, got %v", startStream.events[0])
	}

	// Verify Exit event
	lastIdx := len(startStream.events) - 1
	exitEvent := startStream.events[lastIdx].GetExit()
	if exitEvent == nil || exitEvent.GetExitCode() != 0 {
		t.Errorf("expected ExitEvent with exit code 0, got %v", startStream.events[lastIdx])
	}
}

func TestProcessServerSignal(t *testing.T) {
	tempDir := t.TempDir()
	registry := processmanager.NewProcessRegistry()
	procServer := NewProcessServer(tempDir, registry)

	startStream := &mockStartServer{}

	go func() {
		_ = procServer.Start(&processv1.StartRequest{
			Config: &processv1.ProcessConfig{
				Command: []string{"sleep", "10"},
			},
		}, startStream)
	}()

	// Wait for process to launch and register
	var pid int32
	for i := 0; i < 50; i++ {
		time.Sleep(10 * time.Millisecond)
		active := registry.ListActive()
		if len(active) > 0 {
			pid = active[0].ID
			break
		}
	}

	if pid <= 0 {
		t.Fatalf("process failed to register within timeout")
	}

	// Send SIGKILL signal
	_, err := procServer.SendSignal(context.Background(), &processv1.SendSignalRequest{
		ProcessId: pid,
		Signal:    processv1.Signal_SIGNAL_SIGKILL,
	})
	if err != nil {
		t.Fatalf("SendSignal failed: %v", err)
	}
}

File: clients/typescript/agentic-sandbox-client/src/__tests__/process.test.ts
// Copyright 2026 The Kubernetes Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as http2 from "node:http2";
import type * as net from "node:net";
import { create } from "@bufbuild/protobuf";
import { Code, ConnectError, type ConnectRouter } from "@connectrpc/connect";
import { connectNodeAdapter } from "@connectrpc/connect-node";
import { afterEach, describe, expect, it } from "vitest";
import {
  ExecuteResponseSchema,
  type ProcessConfig,
  ProcessService,
} from "../_proto/process/v1/process_pb.js";
import {
  SandboxClosedError,
  SandboxConnectionError,
  SandboxdRpcError,
} from "../exceptions.js";
import { ProcessClient } from "../process.js";

/**
 * `disruptFirstExecute` severs the transport on the first Execute request
 * instead of letting `routes` handle it, before any response is sent — this
 * is what a real mid-call disconnect looks like (server-side session
 * teardown, or a raw socket reset), as opposed to an application-level gRPC
 * error the router returns intentionally.
 */
async function startH2Server(
  routes: (router: ConnectRouter) => void,
  opts?: { disruptFirstExecute?: "session" | "socket" },
): Promise<{ baseUrl: string; close(): Promise<void> }> {
  const inner = connectNodeAdapter({ routes });
  let triggered = false;
  let lastRawSocket: net.Socket | undefined;
  const server = http2.createServer((req, res) => {
    if (
      !triggered &&
      opts?.disruptFirstExecute &&
      req.url === "/process.v1.ProcessService/Execute"
    ) {
      triggered = true;
      if (opts.disruptFirstExecute === "session") {
        req.stream.session?.destroy();
      } else {
        // req.stream.session.socket is a Proxy that throws
        // ERR_HTTP2_NO_SOCKET_MANIPULATION on destroy(); use the raw socket
        // captured via the server's "connection" event instead, and
        // resetAndDestroy() so the client observes an actual ECONNRESET
        // rather than a clean FIN.
        lastRawSocket?.resetAndDestroy();
      }
      return;
    }
    inner(req, res);
  });
  server.on("connection", (socket: net.Socket) => {
    lastRawSocket = socket;
  });
  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", resolve));
  const addr = server.address();
  if (!addr || typeof addr === "string") throw new Error("failed to bind");
  return {
    baseUrl: `http://127.0.0.1:${addr.port}`,
    close: () => new Promise((resolve) => server.close(() => resolve())),
  };
}

let activeServer: Awaited<ReturnType<typeof startH2Server>> | undefined;
let activeClient: ProcessClient | undefined;

afterEach(async () => {
  activeClient?.abort();
  activeClient = undefined;
  await activeServer?.close();
  activeServer = undefined;
});

describe("ProcessClient.run", () => {
  it("sends the argv as-is (no shell wrapping) and returns stdout/stderr/exitCode", async () => {
    let received: ProcessConfig | undefined;
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: (req) => {
          received = req.config;
          return create(ExecuteResponseSchema, {
            exitCode: 0,
            stdout: new TextEncoder().encode("hello world\n"),
            stderr: new Uint8Array(),
          });
        },
      });
    });
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    const result = await activeClient.run(
      { command: ["echo", "hello world"] },
      30_000,
      new AbortController().signal,
    );
    expect(result).toEqual({
      exitCode: 0,
      stdout: "hello world\n",
      stderr: "",
    });
    expect(received?.command).toEqual(["echo", "hello world"]);
    expect(received?.envVars).toEqual({});
    expect(received?.cwd).toBeUndefined();
  });

  it("passes env and cwd through to ProcessConfig", async () => {
    let received: ProcessConfig | undefined;
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: (req) => {
          received = req.config;
          return create(ExecuteResponseSchema, { exitCode: 0 });
        },
      });
    });
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    await activeClient.run(
      { command: ["env"], env: { FOO: "bar", EMPTY: "" }, cwd: "work/dir" },
      30_000,
      new AbortController().signal,
    );
    expect(received?.command).toEqual(["env"]);
    expect(received?.envVars).toEqual({ FOO: "bar", EMPTY: "" });
    expect(received?.cwd).toBe("work/dir");
  });

  it("returns a non-zero exit code as a normal result, not a thrown error", async () => {
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: () =>
          create(ExecuteResponseSchema, {
            exitCode: 3,
            stdout: new Uint8Array(),
            stderr: new TextEncoder().encode("failed\n"),
          }),
      });
    });
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    const result = await activeClient.run(
      { command: ["false"] },
      30_000,
      new AbortController().signal,
    );
    expect(result.exitCode).toBe(3);
    expect(result.stderr).toBe("failed\n");
  });

  it("maps sandboxd's NOT_FOUND for a missing executable to SandboxdRpcError", async () => {
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: () => {
          throw new ConnectError(
            "failed to execute command: command or path not found",
            Code.NotFound,
          );
        },
      });
    });
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    await expect(
      activeClient.run(
        { command: ["definitely-not-a-real-command"] },
        30_000,
        new AbortController().signal,
      ),
    ).rejects.toSatisfy(
      (err: unknown) =>
        err instanceof SandboxdRpcError && err.code === "not_found",
    );
  });

  it("maps a response exceeding readMaxBytes to resource_exhausted", async () => {
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: () =>
          create(ExecuteResponseSchema, {
            exitCode: 0,
            stdout: new Uint8Array(64 * 1024),
            stderr: new Uint8Array(),
          }),
      });
    });
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      // Small enough that the real decoded message size trips the transport's
      // own wire-level limit — this exercises the actual gRPC/h2 stack, not a
      // client-side pre-check.
      maxCommandOutputSize: 16,
    });
    await expect(
      activeClient.run(
        { command: ["big"] },
        30_000,
        new AbortController().signal,
      ),
    ).rejects.toMatchObject({
      code: "resource_exhausted",
    } satisfies Partial<SandboxdRpcError>);
  });

  it("lets one cancelled concurrent call not affect the other", async () => {
    let releaseSlow: (() => void) | undefined;
    const slowGate = new Promise<void>((resolve) => {
      releaseSlow = resolve;
    });
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: async (req) => {
          if (req.config?.command?.[0] === "slow") {
            await slowGate;
          }
          return create(ExecuteResponseSchema, {
            exitCode: 0,
            stdout: new TextEncoder().encode(req.config?.command?.[0] ?? ""),
            stderr: new Uint8Array(),
          });
        },
      });
    });
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });

    const cancelController = new AbortController();
    const slowRun = activeClient.run(
      { command: ["slow"] },
      30_000,
      cancelController.signal,
    );
    const fastRun = activeClient.run(
      { command: ["fast"] },
      30_000,
      new AbortController().signal,
    );

    cancelController.abort(new Error("caller cancelled"));
    // Must reject with the caller's own abort reason, not get reclassified
    // as a transport failure just because Connect reports it as Canceled.
    await expect(slowRun).rejects.toThrow("caller cancelled");
    await expect(fastRun).resolves.toEqual({
      exitCode: 0,
      stdout: "fast",
      stderr: "",
    });
    releaseSlow?.();
  });

  it("classifies a server-side HTTP/2 session teardown as a connection error", async () => {
    activeServer = await startH2Server(
      (router) => {
        router.service(ProcessService, {
          execute: () => create(ExecuteResponseSchema, {}),
        });
      },
      { disruptFirstExecute: "session" },
    );
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    // Asserts the actual Connect code that reached classifyError(), not just
    // the outcome — this is what makes the test a genuine regression guard:
    // Code.Canceled must be classified alongside Unavailable, not fall
    // through to SandboxdRpcError.
    await expect(
      activeClient.run(
        { command: ["echo", "hi"] },
        30_000,
        new AbortController().signal,
      ),
    ).rejects.toSatisfy(
      (err: unknown) =>
        err instanceof SandboxConnectionError &&
        err.kind === "socket" &&
        err.cause instanceof ConnectError &&
        err.cause.code === Code.Canceled,
    );
  });

  it("classifies a server-side raw socket reset (ECONNRESET) as a connection error", async () => {
    activeServer = await startH2Server(
      (router) => {
        router.service(ProcessService, {
          execute: () => create(ExecuteResponseSchema, {}),
        });
      },
      { disruptFirstExecute: "socket" },
    );
    activeClient = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    await expect(
      activeClient.run(
        { command: ["echo", "hi"] },
        30_000,
        new AbortController().signal,
      ),
    ).rejects.toSatisfy(
      (err: unknown) =>
        err instanceof SandboxConnectionError &&
        err.kind === "socket" &&
        err.cause instanceof ConnectError &&
        err.cause.code === Code.Aborted,
    );
  });

  it("throws SandboxClosedError for any call after abort()", async () => {
    activeServer = await startH2Server((router) => {
      router.service(ProcessService, {
        execute: () =>
          create(ExecuteResponseSchema, {
            exitCode: 0,
            stdout: new Uint8Array(),
            stderr: new Uint8Array(),
          }),
      });
    });
    const client = new ProcessClient({
      grpcBaseUrl: activeServer.baseUrl,
      maxCommandOutputSize: 1024 * 1024,
    });
    client.abort();
    await expect(
      client.run(
        { command: ["echo", "hi"] },
        30_000,
        new AbortController().signal,
      ),
    ).rejects.toBeInstanceOf(SandboxClosedError);
  });
});
